File: Lua/merge_sort.lua
function merge(arr, left, middle, right)
    local n1 = middle - left + 1
    local n2 = right - middle
    local left_arr = {}
    local right_arr = {}

    for i = 1, n1 do
        left_arr[i] = arr[left + i - 1]
    end
    for i = 1, n2 do
        right_arr[i] = arr[middle + i]
    end

    local i, j, k = 1, 1, left
    while i <= n1 and j <= n2 do
        if left_arr[i] <= right_arr[j] then
            arr[k] = left_arr[i]
            i = i + 1
        else
            arr[k] = right_arr[j]
            j = j + 1
        end
        k = k + 1
    end

    while i <= n1 do
        arr[k] = left_arr[i]
        i = i + 1
        k = k + 1
    end

    while j <= n2 do
        arr[k] = right_arr[j]
        j = j + 1
        k = k + 1
    end
end

function merge_sort(arr, left, right)
    if left < right then
        local middle = math.floor((left + right) / 2)
        merge_sort(arr, left, middle)
        merge_sort(arr, middle + 1, right)
        merge(arr, left, middle, right)
    end
end

local arr = {5, 1, 4, 2, 8}
merge_sort(arr, 1, #arr)
print(table.concat(arr, " "))

File: C#/MergeSort.cs
using System;

class MergeSort
{
    public static void Sort(int[] arr, int left, int right)
    {
        if (left < right)
        {
            int mid = left + (right - left) / 2;
            Sort(arr, left, mid);
            Sort(arr, mid + 1, right);
            Merge(arr, left, mid, right);
        }
    }

    private static void Merge(int[] arr, int left, int mid, int right)
    {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] leftArr = new int[n1];
        int[] rightArr = new int[n2];

        for (int i = 0; i < n1; i++)
            leftArr[i] = arr[left + i];

        for (int j = 0; j < n2; j++)
            rightArr[j] = arr[mid + 1 + j];

        int i1 = 0, i2 = 0, k = left;
        while (i1 < n1 && i2 < n2)
            arr[k++] = leftArr[i1] <= rightArr[i2] ? leftArr[i1++] : rightArr[i2++];

        while (i1 < n1) arr[k++] = leftArr[i1++];
        while (i2 < n2) arr[k++] = rightArr[i2++];
    }

    public static void Main()
    {
        int[] arr = { 12, 11, 13, 5, 6, 7 };
        Sort(arr, 0, arr.Length - 1);
        Console.WriteLine(string.Join(" ", arr));
    }
}
